entity-service-impl/src/main/java/org/hypertrace/entity/type/service/v2/EntityTypeServiceImplV2.java

```diff
@@ -0,0 +1,122 @@
+package org.hypertrace.entity.type.service.v2;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.protobuf.ServiceException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import org.hypertrace.core.documentstore.Collection;
+import org.hypertrace.core.documentstore.Datastore;
+import org.hypertrace.core.documentstore.Document;
+import org.hypertrace.core.documentstore.Filter;
+import org.hypertrace.core.documentstore.Filter.Op;
+import org.hypertrace.core.documentstore.Query;
+import org.hypertrace.core.documentstore.SingleValueKey;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+import org.hypertrace.entity.service.constants.EntityCollectionConstants;
+import org.hypertrace.entity.service.constants.EntityServiceConstants;
+import org.hypertrace.entity.service.util.TenantUtils;
+import org.hypertrace.entity.type.service.v2.EntityTypeServiceV2Grpc.EntityTypeServiceV2ImplBase;
+import org.hypertrace.entity.type.service.v2.model.EntityTypeDocument;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class EntityTypeServiceImplV2 extends EntityTypeServiceV2ImplBase {
+  private static final Logger LOGGER = LoggerFactory.getLogger(EntityTypeServiceImplV2.class);
+
+  private final Collection entityTypeCollection;
+
+  public EntityTypeServiceImplV2(Datastore datastore) {
+    this.entityTypeCollection =
+        datastore.getCollection(EntityCollectionConstants.ENTITY_TYPES_V2_COLLECTION);
+  }
+
+  @Override
+  public void upsertEntityType(org.hypertrace.entity.type.service.v2.EntityType request,
+      io.grpc.stub.StreamObserver<org.hypertrace.entity.type.service.v2.EntityType> responseObserver) {
+    Optional<String> tenantId = RequestContext.CURRENT.get().getTenantId();
+    if (tenantId.isEmpty()) {
+      responseObserver.onError(new ServiceException("Tenant id is missing in the request."));
+      return;
+    }
+
+    try {
+      // Set tenant id.
+      EntityTypeDocument document = EntityTypeDocument.fromProto(request);
+      document.setTenantId(tenantId.get());
+      entityTypeCollection.upsert(new SingleValueKey(tenantId.get(), request.getName()), document);
```

Comment:
I would expect upsert to return the upserted value rather than the requested value - what's the behavior if this entity type already exists with different values? Or a superset of values?

Reply:
Done